compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -259,10 +259,11 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
     ) -> ImplSourceAutoImplData<PredicateObligation<'tcx>> {
         debug!(?obligation, ?trait_def_id, "confirm_auto_impl_candidate");
 
-        let types = obligation.predicate.map_bound(|inner| {
-            let self_ty = self.infcx.shallow_resolve(inner.self_ty());
-            self.constituent_types_for_ty(self_ty)
-        });
+        let self_ty = self.infcx.shallow_resolve(obligation.predicate.skip_binder().self_ty());
+        // Really, this is only wrapped in a `Binder` because of `GeneratorWitness`
+        // In that case, we don't expect outer binders
+        let constituent_types = self.constituent_types_for_ty(self_ty);
+        let types = ty::Binder::fuse(constituent_types, obligation.predicate, |types, _| types);
```

Comment:
```suggestion
        let types = obligation.predicate.rebind(constituent_types);
```
?

Reply:
No - this is a weird case. See https://github.com/rust-lang/rust/pull/80106/files#diff-34d5893dd95fe09f9a0fd3341efacd1c21853bb34ba29d9d79bb9af26bb8a0a0R1697-R1700. The `GeneratorWitness` `types` are under a *different* set of binders (at least in theory).

Reply:
my understanding was that we expect the two binders to be the same for `fuse` as well, so these two expressions should be identical, even after your bound vars work. Do we recompute the binders in `fuse` or union them?

Reply:
So, `fuse` is a little weird. In Chalk, `Binder::fuse` takes a `Binder<Binder<T>>` and will do the shifting of the inner vars. In rustc, `fuse` is a more abstract operation that takes two bound types and combines them via the supplied function.

In #76814, I take this to mean that the two binders are *compatible* but not necessarily the *same*. In practice, that means that the out binders are identical to the longer binders. Also there, *this* ends up being the only place that needs `fuse`, so maybe the right thing here is to just `ty::Binder::bind(types.skip_binder())` and in that PR, we can be explicit about picking the binders that aren't `len == 0`

Reply:
> In practice, that means that the out binders are identical to the longer binders

so you expect one of the two binders to be a subset of the other one?

Reply:
Err, I guess that's not *quite* right on master. On master, I guess technically everything is bound after.

As we start to assign indices to bound vars, the guarantees we require start to get stronger. Eventually, yes, we expect one to be a subset of the other.

I'm actually going to just remove `fuse` in the binder refactor PR in favor of the explicit stronger guarantee here.

Reply:
I am still a bit confused here :thinking:

Afaict `constituent_types` only contains bound vars inside of a `GeneratorWitness` as all other types are wrapped with `Binder::dummy`. Doesn't this mean that we don't have to rebind/fuse at all here

Reply:
Oh uh, I guess you're right. I'm going to add an assert that `self_ty` doesn't have escaping vars and just remove the fuse.